Feat register and misc

elkcert opened the pull request.

elkcert committed:
feat(back): Automatic DTO mapping infrastructure

Copied from https://auth0.com/blog/automatically-mapping-dto-to-entity-on-spring-boot-apis/

elkcert committed:
feat(back): Throw if entity not found in mapper

elkcert committed:
refactor(back): Generalize controller exception handling

elkcert committed:
refactor(back): Use dto mapper for register endpoint

elkcert committed:
feat(front): Beginning of register

elkcert committed:
refactor(back): User implements UserDetails

elkcert committed:
feat(back): Port BadValidation from previous project

elkcert committed:
fix(back): Ambiguous exception handlers

elkcert committed:
fix(back): Default role for user

elkcert committed:
feat(front): Register with validation

elkcert committed:
refactor(front): Login with validation

magley commented on back/sssscs/src/main/java/com/ib/common/EntityNotFoundException.java:
```diff
@@ -0,0 +1,8 @@
+
+public class EntityNotFoundException extends EntityException {
+
+    public EntityNotFoundException(Class entityType, Long id) {
```
Eclipse is complaining that this should be parameterized, like ```Class<?>```. Is there a reason why it's just `Class` here (and in all the other files)?

elkcert replied:
Missed it since IntelliJ only shows warnings when you're in the specific file, not project-wide like Eclipse
https://github.com/magley/sssscs/pull/2/commits/0c2faad0e93d9f57f4fe555dcd8e400fcb1ac2f7

magley commented on back/sssscs/src/main/java/com/ib/user/User.java:
```diff
@@ -19,12 +25,48 @@
+public class User implements UserDetails {
+
+	public enum Role {
+		FOO,
```
We should add the actual roles right now, there's only two roles anyway (not counting the logged-out user).

elkcert replied:
https://github.com/magley/sssscs/pull/2/commits/d58052c8b2394e9fdccabc86620a4121b5472ab4

magley commented on back/sssscs/src/main/java/com/ib/util/BadValidation.java:
```diff
@@ -0,0 +1,21 @@
+
+    @JsonInclude(value = Include.NON_NULL)
+    private Long timestamp;
+    private String message;
```
Why is this field excluded from json?

elkcert replied:
Suppose it's an optimization so we don't serialize more than what is necessary. `error` and `path` are excluded as well. @swopl wrote this, so he might have more to say.

swopl replied:
They are excluded if null. I think i did this just to get a prettier json instead of having path: null. Idk, I think they all have non_null anyways

swopl replied:
I remembered. On the previous project some needed to only have message shown in json for exact matching in those bad python tests we had. You can safely do away with these now

magley commented on back/sssscs/src/main/java/com/ib/util/DTOModelMapper.java:
```diff
@@ -0,0 +1,89 @@
+    }
+
+    private Object getEntityId(@NotNull Object dto) {
+        for (Field field : dto.getClass().getDeclaredFields()) {
```
`getDeclaredFields()` could cause problems if we have entities that inherit from other entities (inherited `@Id` field). Will this even be an issue for us; would we even want to have an entity class hierarchy?

elkcert replied:
If it becomes an issue, we could replace it with `getFields()`. Probably just leave it like this for now.

magley commented on front/sssscs/src/register/Register.tsx:
```diff
@@ -0,0 +1,90 @@
+
+const TryRegister = async (data: FieldValues) => {
+    const dto: UserCreateDto = {
+        'email': data['email'],
```
Use 
```
email: data['email']
```
instead of 
```
'email': data['email']
```

elkcert replied:
https://github.com/magley/sssscs/pull/2/commits/4fd34a72e41fb190e728d2d2a9cd4284826b6c08

magley commented on front/sssscs/src/register/Register.tsx:
```diff
@@ -0,0 +1,90 @@
+        });
+}
+
+const Register = () => {
```
We should agree on whether to add type annotations to _all_ components or _none_ of them. I think the syntax is cubersome:
```ts
const ComponentName: React.FC<{}> = () => {
```
but once our components become complex enough to need state or props, it can help with readability.

@swopl Thoughts?

swopl replied:
Do we need the types for components? Are they useful anywhere? I prefer the way without them

elkcert replied:
https://github.com/magley/sssscs/pull/2/commits/4fd34a72e41fb190e728d2d2a9cd4284826b6c08

magley commented on front/sssscs/src/login/Login.tsx:
```diff
@@ -21,31 +25,33 @@ const TryLogin = async (data: UserLoginDto) => {
+                sx={{mb: 2}}
+                label="Email"
+                fullWidth
+                {...register('email', { required: 'Email is required' })}
```
Should we add an email regex rule here as well?

elkcert replied:
I was thinking we could leave it out of the Login form since we're only using it to retrieve data, and not for posting like in the Register form.

magley commented on front/sssscs/src/register/Register.tsx:
```diff
@@ -0,0 +1,90 @@
+                        message: 'Not a valid email'
+                    }
+                })}
+                error={!!errors['email']}
```
How would we go about adding an error message if the email is already taken i.e. after a backend response?

elkcert replied:
Maybe something like 
```Typescript
error={!!errors['email'] || isEmailTaken}
helperText={errors['email'] ? errors['email'].message.toString() : isEmailTaken ? 'Email taken' : ''}
```
Looks a little ugly, could maybe be refined

magley commented on back/sssscs/src/main/java/com/ib/user/User.java:
```diff
@@ -19,12 +25,48 @@
+	@Id
 	@GeneratedValue(strategy = GenerationType.IDENTITY)
 	private Long id;
 	private String email;
```
I think that the phone number is one of the required `User` fields as well.

elkcert replied:
https://github.com/magley/sssscs/pull/2/commits/9211a0d36efcb18edf5741710eb7bc8e759fce5f

magley replied:
Aside from the missing extra validation rules for the phone number, there should generally be a stronger validation of the registration form on the frontend.
For example, the example JSON body for the postman of the endpoint is only 4 fields long, but it passes the registration. Everything should _at least_ be required.

elkcert replied:
> For example, the example JSON body for the postman of the endpoint is only 4 fields long, but it passes the registration. Everything should _at least_ be required.

This sounds like a backend validation problem, no? I didn't think of adding backend validation in this branch, but why not. https://github.com/magley/sssscs/pull/2/commits/c699227fb2beef436474d263a3d73676d35c36dc

magley replied:
Well actually I did mean validation on the _frontend_, having each text field marked as required, but no harm in doing backend validation now, I guess.

magley reviewed (changes requested):
Aside from the comments above, have you thought about which token mechanism we'll use (jwt, cookie, etc.)?

elkcert committed:
refactor(back): Warning cleanup

elkcert committed:
feat(back): User roles

elkcert committed:
refactor(front): Simplify some syntax

elkcert committed:
feat(all): User phone number

elkcert commented:
> Aside from the comments above, have you thought about which token mechanism we'll use (jwt, cookie, etc.)?

I was thinking of just using JWT for now.

elkcert committed:
feat(all): Stronger validation

swopl commented on back/sssscs/src/main/java/com/ib/config/ControllerAdvisor.java:
```diff
@@ -1,20 +1,45 @@
+
+	@ExceptionHandler({ MethodArgumentNotValidException.class })
+	public ResponseEntity<BadValidation> handleConstraintViolationException(MethodArgumentNotValidException e,
+																			   HttpServletRequest req) {
```
Formatting? Maybe github just messed it up for me

elkcert replied:
IntelliJ formatted it like that. https://github.com/magley/sssscs/pull/2/commits/ceec7bb98e542c2181a214165dc4615668b3ce3a

swopl commented on back/sssscs/src/main/java/com/ib/config/ServerConfig.java:
```diff
@@ -1,15 +1,38 @@
+	@Override
+	public void addArgumentResolvers(List<HandlerMethodArgumentResolver> argumentResolvers) {
+		ObjectMapper objectMapper = Jackson2ObjectMapperBuilder.json().applicationContext(this.applicationContext).build();
+		argumentResolvers.add(new DTOModelMapper(objectMapper, entityManager));
```
Do you need to build a new ObjectMapper here? Is it problematic to use an autowired one, or does it not even exist yet?

swopl commented on back/sssscs/src/main/java/com/ib/util/DTOModelMapper.java:
```diff
@@ -0,0 +1,90 @@
+        } else {
+            Class<?> entityType = parameter.getParameterType();
+            Object persistedObject = entityManager.find(entityType, id);
+            if (persistedObject == null) {
+                throw new EntityNotFoundException(entityType, (Long) id);
+            }
```
Have you tested this to see if it works? Automatic tests would be nice for this at least

swopl commented on back/sssscs/src/main/java/com/ib/common/EntityNotFoundException.java:
```diff
@@ -0,0 +1,8 @@
+    public EntityNotFoundException(Class<?> entityType, Long id) {
+        super(String.format("%s with id %d not found", entityType.getName(), id));
+    }
+}
```
What is the difference between common and util?

elkcert replied:
Probably don't need them to be different packages. https://github.com/magley/sssscs/pull/2/commits/08eb9236237c00124882f6de8468a31e375586e5

swopl commented on front/sssscs/src/register/Register.tsx:
```diff
@@ -0,0 +1,129 @@
+import { FieldValues, useForm } from "react-hook-form";
+
+const TryRegister = async (data: FieldValues) => {
+    const dto: UserCreateDto = {
```
Why is this function name capitalized? I thought this would be a component on first read

elkcert replied:
https://github.com/magley/sssscs/pull/2/commits/ceec7bb98e542c2181a214165dc4615668b3ce3a

swopl commented on back/sssscs/src/main/java/com/ib/user/UserController.java:
```diff
@@ -28,9 +22,10 @@ public class UserController {
-		User user = userService.register(dto);
-		return ResponseEntity.ok(user);
+	public ResponseEntity<User> register(@DTO(UserCreateDto.class) User user) {
+		// FIXME: Return something other than user. Also, why do we return 500 with
+		//        "Failed to write request" now that User implements UserDetails?
+		return ResponseEntity.ok(userService.register(user));
```
Didn't return 500 for me?

elkcert replied:
Obsolete comment. https://github.com/magley/sssscs/pull/2/commits/ceec7bb98e542c2181a214165dc4615668b3ce3a

swopl commented on front/sssscs/src/login/Login.tsx:
```diff
@@ -1,15 +1,19 @@
  */
-const TryLogin = async (data: UserLoginDto) => {
-    LoginService.login(data)
+const TryLogin = async (data: FieldValues) => {
+    const dto: UserLoginDto = {
```
Why is this function name capitalized? I thought this would be a component on first read
Usually in React components start with capital letters and hooks start with use[A-Z]

elkcert replied:
https://github.com/magley/sssscs/pull/2/commits/ceec7bb98e542c2181a214165dc4615668b3ce3a

swopl commented on front/sssscs/src/register/Register.tsx:
```diff
@@ -0,0 +1,129 @@
+                    }
+                })}
+                error={!!errors['password']}
+                helperText={errors['password']?.message?.toString()}
```
Add password pattern. We need one for secure password. I think this is even one of the project requests

elkcert replied:
https://github.com/magley/sssscs/pull/2/commits/199d1c1fe63bb093c3782136221c92c08ee69429

elkcert committed:
feat(front): Further improvements to validation

- Confirm password validation now works properly
- Validations trigger on all changes
- Required fields are clearly marked with an asterisk (*)

elkcert committed:
refactor(all): Remove a bunch of silliness

elkcert committed:
refactor(back): Remove common package

elkcert committed:
feat(all): Password regex

elkcert committed:
chore(back): todos
